blocks/log.py

```diff
@@ -1,274 +1,117 @@
-        value = self._fetch_record(time, key)
-        if value is not None:
-            return value
-        return self.get_default_value(key)
-
-    @abstractmethod
-    def _fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        The implementation method to be overridden.
+    def current_entry(self):
+        return self[self.status['iterations_done']]
 
-        """
-        pass
+    @property
+    def previous_entry(self):
+        return self[self.status['iterations_done'] - 1]
 
-    def __getitem__(self, time):
-        self._check_time(time)
-        return TrainingLogRow(self, time)
 
-    @property
-    def current_row(self):
-        return self[self.status.iterations_done]
+class TrainingStatus(Mapping):
+    def __init__(self, exclude=None):
+        self._status = {'iterations_done': 0, 'epochs_done': 0}
+        self.exclude = [] if exclude is None else exclude
 
-    @property
-    def previous_row(self):
-        return self[self.status.iterations_done - 1]
+    def __getitem__(self, key):
+        return self._status[key]
 
-    @property
-    def last_epoch_row(self):
-        return self[self.status._epoch_ends[-1]]
+    def __setitem__(self, key, value):
+        self._status[key] = value
 
-    @abstractmethod
     def __iter__(self):
-        """Returns an iterator over time-key-value triples of the log."""
-        pass
+        return (key for key in self._status if key not in self.exclude)
 
-    @abstractmethod
-    def get_row_iterator(self, time):
-        """Returns an iterator over key-value pairs of a row of the log."""
-        pass
+    def __len__(self):
+        return len(self._status)
 
-    def _check_time(self, time):
-        if not isinstance(time, int) or time < 0:
-            raise ValueError("time must be a positive integer")
 
-    def to_dataframe(self):
-        """Convert a log into a :class:`.DataFrame`."""
-        if not PANDAS_AVAILABLE:
-            raise ImportError("The pandas library is not found. You can"
-                              " install it with pip.")
-        return self._to_dataframe()
-
-    def _to_dataframe(self):
-        raise NotImplementedError()
+class MongoTrainingLog(Mapping):
+    def __init__(self):
+        self.client = MongoClient()
+        self.db = self.client.blocks_log
+        self.entries = self.db.entries
 
+    def __getitem__(self, key):
+        return MongoEntry(self, key)
 
-class TrainingStatus(AbstractTrainingStatus):
-    """A simple training status."""
     def __iter__(self):
-        for attr, value in self.__dict__.items():
-            if not attr.startswith("__"):
-                yield attr, value
-
-
-class TrainingLog(AbstractTrainingLog):
-    """A simple training log storing information in main memory."""
-    def __init__(self):
-        self._storage = defaultdict(dict)
-        self._default_values = {}
-        self._status = TrainingStatus()
+        return map(methodcaller('pop', '_id'),
```

Comment:
I was scared about the overhead, giving that this could be used to request 100,000s of log entries in a query. I'll just replace it with `imap` from `picklable_itertools` for now and if it turns out to be slow I'll look at it again.